test/integration/storageversionmigrator/storageversionmigrator_test.go

```diff
@@ -0,0 +1,111 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageversionmigrator
+
+import (
+	"bytes"
+	"context"
+	"testing"
+	"time"
+
+	"k8s.io/klog/v2/ktesting"
+	"k8s.io/kubernetes/pkg/features"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	encryptionconfigcontroller "k8s.io/apiserver/pkg/server/options/encryptionconfig/controller"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+)
+
+// TestStorageVersionMigration is an integration test that verifies storage version migration works.
+// This test asserts following scenarios:
+// 1. Start API server with encr at rest and hot reload of encryption config enabled
+// 2. Create a secret
+// 3. Update encryption config file to add a new key as write key
+// 4. Perform Storage Version Migration for secrets
+// 5. Verify that the secret is migrated to use the new key
+// 6. Verify that the secret is updated with a new resource version
+// 7. Perform another Storage Version Migration for secrets
+// 8. Verify that the resource version of the secret is not updated. i.e. it was a no-op update
+func TestStorageVersionMigration(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionMigrator, true)()
+	// this makes the test super responsive. It's set to a default of 1 minute.
+	encryptionconfigcontroller.EncryptionConfigFileChangePollDuration = time.Millisecond
+
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	svmTest := svmSetup(ctx, t)
+
+	// ToDo: try to test with 1000 secrets
+	secret, err := svmTest.createSecret(ctx, t, secretName, secretNamespace)
+	if err != nil {
+		t.Fatalf("Failed to create secret: %v", err)
+	}
+
+	metricBeforeUpdate := svmTest.getAutomaticReloadSuccessTotal(ctx, t)
+	svmTest.updateFile(t, svmTest.filePathForEncryptionConfig, encryptionConfigFileName, []byte(resources["updatedEncryptionConfig"]))
+	if !svmTest.isEncryptionConfigFileUpdated(ctx, t, metricBeforeUpdate) {
+		t.Fatalf("Failed to update encryption config file")
+	}
+
+	svm, err := svmTest.createSVMResource(ctx, t, svmName)
+	if err != nil {
+		t.Fatalf("Failed to create SVM resource: %v", err)
+	}
+	if !svmTest.isResourceMigrated(ctx, t, svm.Name, secret.Name, secret.Namespace, 1) {
+		t.Fatalf("Failed to migrate resource %s/%s", secret.Namespace, secret.Name)
+	}
+
+	wantPrefix := "k8s:enc:aescbc:v1:key2"
+	etcdSecret, err := svmTest.getRawSecretFromETCD(t, secret.Name, secret.Namespace)
+	if err != nil {
+		t.Fatalf("Failed to get secret from etcd: %v", err)
+	}
+	// assert that secret is prefixed with the new key
+	if !bytes.HasPrefix(etcdSecret, []byte(wantPrefix)) {
```

Comment:
I don't understand the purpose of this check for the test.

Reply:
Since in this case we are using SVM with KMS I thought it would be a good idea to validate etcd prefix to make sure it's using the new key (and hence another check that SVM has worked). Should I remove it?

Reply:
I don't object, but we need to be sure we cover a non-kms case as Joe mentioned above.